doc/compiler/optimizer/IntroReadLogFile.md

```diff
@@ -0,0 +1,1188 @@
+
+`"Pre Instruction Selection Trees"` are the last set of trees before getting into the code gen.
+
+### 7.2 Instruction Selection
+
+The first step of generating code is `“Performing Instruction Selection”`. It selects specific instructions that need
+to generate on a platform. There are three things essential to generate code for a specific platform.
+1.  Instructions for the opcode
+2.  Registers
+3.  Encoding (the way the instructions encoded in the generated code cache)
+
+```
+============================================================
+; Live regs: GPR=2 FPR=0 VRF=0 {&GPR_0033, GPR_0032}
+------------------------------
+ n642n    (  0)  ificmpeq --> block_2 BBStart at n5n ()                                               [0x7fef25795860] bci=[-1,0,12] rc=0 vc=4181 vn=- li=53 udi=- nc=3 flg=0x20
+ n640n    (  1)    iload  0x7fef247330cc[#479  Static] [flags 0x10303 0x8040 ] (cannotOverflow )      [0x7fef257957c0] bci=[-1,0,12] rc=1 vc=4181 vn=- li=53 udi=- nc=0 flg=0x1000
+ n641n    (  1)    iconst -1 (X!=0 X<=0 )                                                             [0x7fef25795810] bci=[-1,0,12] rc=1 vc=4181 vn=- li=53 udi=- nc=0 flg=0x204
+ n4813n   (  1)    GlRegDeps                                                                          [0x7fef26427020] bci=[-1,0,12] rc=1 vc=4181 vn=- li=53 udi=- nc=2
+ n4811n   (  5)      ==>lRegLoad (in GPR_0032) (cannotOverflow SeenRealReference )
+ n4812n   (  3)      ==>aRegLoad (in &GPR_0033)
+------------------------------
+------------------------------
+ n642n    (  0)  ificmpeq --> block_2 BBStart at n5n ()                                               [0x7fef25795860] bci=[-1,0,12] rc=0 vc=4181 vn=- li=53 udi=- nc=3 flg=0x20
+ n640n    (  0)    iload  0x7fef247330cc[#479  Static] [flags 0x10303 0x8040 ] (cannotOverflow )      [0x7fef257957c0] bci=[-1,0,12] rc=0 vc=4181 vn=- li=53 udi=- nc=0 flg=0x1000
+ n641n    (  0)    iconst -1 (X!=0 X<=0 )                                                             [0x7fef25795810] bci=[-1,0,12] rc=0 vc=4181 vn=- li=53 udi=- nc=0 flg=0x204
+ n4813n   (  1)    GlRegDeps                                                                          [0x7fef26427020] bci=[-1,0,12] rc=1 vc=4181 vn=- li=53 udi=- nc=2
+ n4811n   (  4)      ==>lRegLoad (in GPR_0032) (cannotOverflow SeenRealReference )
+ n4812n   (  2)      ==>aRegLoad (in &GPR_0033)
+------------------------------
+
+ [0x7fef12ad29f0]	cmp	dword ptr [$0x00007fef247330cc], 0xffffffffffffffff	# CMP4MemImms, SymRef  0x7fef247330cc[#627  Static] [flags 0x10303 0x8040 ]
+ [0x7fef12ad2db0]	assocreg                        # assocreg
+	POST: [&GPR_0033 : eax] [GPR_0032 : esi]
+ [0x7fef12ad2b40]	je	Label L0016			# JE4
+	 PRE: [&GPR_0033 : eax] [GPR_0032 : esi]
+	POST: [&GPR_0033 : eax] [GPR_0032 : esi]
+```
+The generated instruction at `0x7fef12ad29f0` puts a constant value into a memory reference and compare it with `-1`.
```

Comment:
`compare it` -> `compares it`.

I wonder if instead of just diving into specific examples, we might want to first describe with what each of the sections mean (i.e. the two IL Trees, and list of instructions).

Reply:
@joransiu About the three sections shown in Instruction Selection, my understanding is that the first IL trees are the trees that are evaluated to generate the instructions listed below the second IL trees. I'm not too sure of the difference between the first and the second IL trees.

```
 ============================================================
; Live regs: ...
------------------------------
// First ILTrees 
------------------------------
------------------------------
// Second ILTrees
------------------------------
// generated instructions 
```

Reply:
The First IL Trees are pre-evaluation, while the second IL Trees are post-evaluation.  In the latter set, the IL nodes will be tagged with the correspond virtual register, if any.  Also, the reference counts of the various IL nodes will be decremented as appropriate.